rclrs/src/parameter.rs

```diff
@@ -3,3 +3,1299 @@ mod value;
+        node.use_undeclared_parameters()
+            .set("declared_int", 20)
+            .unwrap();
+        let param = node
+            .declare_parameter("declared_int", 30)
+            .mandatory()
+            .unwrap();
+        assert_eq!(param.get(), 20);
+    }
+
+    #[test]
+    fn test_parameter_scope_redeclaring() {
+        let ctx = Context::new([
+            String::from("--ros-args"),
+            String::from("-p"),
+            String::from("declared_int:=10"),
+        ])
+        .unwrap();
+        let node = create_node(&ctx, "param_test_node").unwrap();
+        {
+            // Setting a parameter with an override
+            let param = node
+                .declare_parameter("declared_int", 1)
+                .mandatory()
+                .unwrap();
+            assert_eq!(param.get(), 10);
+            param.set(2).unwrap();
+            assert_eq!(param.get(), 2);
+            // Redeclaring should fail
+            assert!(matches!(
+                node.declare_parameter("declared_int", 1).mandatory(),
+                Err(DeclarationError::AlreadyDeclared)
+            ));
+        }
+        {
+            // Parameter went out of scope, redeclaring should be OK and return command line
+            // override
+            let param = node
+                .declare_parameter("declared_int", 1)
+                .mandatory()
+                .unwrap();
+            assert_eq!(param.get(), 10);
+        }
+        // After a declared parameter went out of scope and was cleared, it should still be
+        // possible to use it as an undeclared parameter, type can now be changed
+        assert!(node
+            .use_undeclared_parameters()
+            .get::<i64>("declared_int")
+            .is_none());
+        node.use_undeclared_parameters()
+            .set("declared_int", 1.0)
+            .unwrap();
+        assert_eq!(
+            node.use_undeclared_parameters().get::<f64>("declared_int"),
+            Some(1.0)
+        );
+    }
+
+    #[test]
+    fn test_parameter_ranges() {
+        let ctx = Context::new([]).unwrap();
+        let node = create_node(&ctx, "param_test_node").unwrap();
+        // Setting invalid ranges should fail
+        let range = ParameterRange {
+            lower: Some(10),
+            upper: Some(-10),
+            step: Some(3),
+        };
+        assert!(matches!(
+            node.declare_parameter("int_param", 5)
+                .range(range)
+                .mandatory(),
+            Err(DeclarationError::InvalidRange)
+        ));
+        let range = ParameterRange {
+            lower: Some(-10),
+            upper: Some(10),
+            step: Some(-1),
+        };
+        assert!(matches!(
+            node.declare_parameter("int_param", 5)
+                .range(range)
+                .mandatory(),
+            Err(DeclarationError::InvalidRange)
+        ));
+        // Setting parameters out of range range should fail
```

Comment:
nit: Duplicate range?

Reply:
Not really, they are all slightly similar but the two previously defined ones are invalid, I guess an alternative could be to declare the first one as mutable and only update single fields but not much would change.

Reply:
I think he means the word "range" is repeated in the documentation.

Reply:
:facepalm: https://github.com/ros2-rust/ros2_rust/pull/332/commits/3e2a7bf2bec3e671a0c3bf877b19166659206928